src/core/net/dns_headers.hpp

```diff
@@ -636,24 +634,104 @@ class Name
+     * @param[in]    aMessage         The message to read the name from. `aMessage.GetOffset()` MUST point to
+     *                                the start of DNS header (this is used to handle compressed names).
      * @param[inout] aOffset          On input, the offset in @p aMessage pointing to the start of the name field.
      *                                On exit (when parsed successfully), @p aOffset is updated to point to the byte
      *                                after the end of name field.
-     * @param[in]    aHeaderOffset    The offset in @p aMessage to the start of the DNS header.
      * @param[out]   aNameBuffer      A pointer to a char array to output the read name as a null-terminated C string.
-     * @param[inout] aNameBufferSize  The maximum number chars in @p aNameBuffer array.
+     * @param[in]    aNameBufferSize  The maximum number chars in @p aNameBuffer array.
      *
      * @retval OT_ERROR_NONE          Successfully read the name, @p aNameBuffer and @p Offset are updated.
      * @retval OT_ERROR_PARSE         Name could not be parsed (invalid format).
      * @retval OT_ERROR_NO_BUFS       Name could not fit in @p aNameBufferSize chars.
      *
      */
-    static otError ReadName(const Message &aMessage,
-                            uint16_t &     aOffset,
-                            uint16_t       aHeaderOffset,
-                            char *         aNameBuffer,
-                            uint16_t       aNameBufferSize);
+    static otError ReadName(const Message &aMessage, uint16_t &aOffset, char *aNameBuffer, uint16_t aNameBufferSize);
+
+    /**
+     * This static method compares a single name label from a message with a given label string.
+     *
+     * This method can be used to compare labels one by one. It checks whether the label read from @p aMessage matches
+     * @p aLabel string.
+     *
+     * Unlike `CompareName()` which requires the labels in the the name string to contain no dot '.' character, this
+     * method allows @p aLabel to include any character.
+     *
+     * @param[in]    aMessage         The message to read the label from to compare. `aMessage.GetOffset()` MUST point
+     *                                to the start of DNS header (this is used to handle compressed names).
+     * @param[inout] aOffset          On input, the offset in @p aMessage pointing to the start of the label to read.
+     *                                On exit and only when label is successfully read and does match @p aLabel,
+     *                                @p aOffset is updated to point to the start of the next label.
+     * @param[in]    aLabel           A pointer to a null terminated string containing the label to compare with.
+
+     * @retval OT_ERROR_NONE          The label from @p aMessage matches @p aLabel. @p aOffset is updated.
+     * @retval OT_ERROR_NOT_FOUND     The label from @p aMessage does not match @p aLabel (note that @p aOffset is not
```

Comment:
This method allows the labels in an encoded name in `message` to be read and compared one by one.

One thing to have in mind (and is different with labels vs full name) is that as we read and compare the labels, the offset can jump back in the message  in the case of compressed name.

I considered different models on how to update/manage the `aOffset` here:
1. If the label does not match, we skip over the label and update the offset to end of current label. But this did not make sense to me:
   - Why would the user want to skip over just a label if it does not match. It makes more sense for user to want try to compare label with maybe something else (in this case offset should remian the same)
   - `aOffset` would not be tracking end of original name field sicne we may have jumped back in message as we read labels one by one.
 2. If the label does not match, we skip over the entire name field. This makes more sense, but implementing it complicates the method and its use:
    - Now we need the user to provide multiple offsets as paramters to this method. The offset to start of the original name field in addition to offset to current label to read (just so that we skip over the original name filed if the label does not match).
    - This makes this method more complicated to use for a situation which may not that common/useful
       - If users want this behavior they can keep track of original offset of name themselve and then use `ParseName()` to parse and skip over it.
3. I decided on this model which made most sense to me. It's simple and practical (i.e, we do not update the offset if the label does not match).

Reply:
Make sense to me. Thanks!